transports/quic/src/transport.rs

```diff
@@ -247,119 +249,123 @@ impl<P: Provider> Transport for GenTransport<P> {
-                            SocketFamily::Ipv6 => SocketAddr::new(Ipv6Addr::UNSPECIFIED.into(), 0),
-                        };
-                        let socket =
-                            UdpSocket::bind(listen_socket_addr).map_err(Self::Error::from)?;
-                        let endpoint_config = self.quinn_config.endpoint_config.clone();
-                        let endpoint = Self::new_endpoint(endpoint_config, None, socket)?;
-
-                        vacant.insert(endpoint.clone());
-                        endpoint
-                    }
-                };
-                dialer
-            }
-            Some(listener) => listener.endpoint.clone(),
-        };
-        let handshake_timeout = self.handshake_timeout;
-        let mut client_config = self.quinn_config.client_config.clone();
-        if version == ProtocolVersion::Draft29 {
-            client_config.version(0xff00_001d);
-        }
-        Ok(Box::pin(async move {
-            // This `"l"` seems necessary because an empty string is an invalid domain
-            // name. While we don't use domain names, the underlying rustls library
-            // is based upon the assumption that we do.
-            let connecting = endpoint
-                .connect_with(client_config, socket_addr, "l")
-                .map_err(ConnectError)?;
-            Connecting::new(connecting, handshake_timeout).await
-        }))
-    }
-
-    fn dial_as_listener(
+    fn dial(
         &mut self,
         addr: Multiaddr,
+        dial_opts: DialOpts,
     ) -> Result<Self::Dial, TransportError<Self::Error>> {
-        let (socket_addr, _version, peer_id) =
+        let (socket_addr, version, peer_id) =
             self.remote_multiaddr_to_socketaddr(addr.clone(), true)?;
-        let peer_id = peer_id.ok_or(TransportError::MultiaddrNotSupported(addr.clone()))?;
-
-        let socket = self
-            .eligible_listener(&socket_addr)
-            .ok_or(TransportError::Other(
-                Error::NoActiveListenerForDialAsListener,
-            ))?
-            .try_clone_socket()
-            .map_err(Self::Error::from)?;
-
-        tracing::debug!("Preparing for hole-punch from {addr}");
-
-        let hole_puncher = hole_puncher::<P>(socket, socket_addr, self.handshake_timeout);
-
-        let (sender, receiver) = oneshot::channel();
-
-        match self.hole_punch_attempts.entry(socket_addr) {
-            Entry::Occupied(mut sender_entry) => {
-                // Stale senders, i.e. from failed hole punches are not removed.
-                // Thus, we can just overwrite a stale sender.
-                if !sender_entry.get().is_canceled() {
-                    return Err(TransportError::Other(Error::HolePunchInProgress(
-                        socket_addr,
-                    )));
+
+        match (dial_opts.role, dial_opts.port_use) {
+            (Endpoint::Dialer, _) | (Endpoint::Listener, PortUse::Reuse) => {
```

Comment:
In the case of `(Endpoint::Dialer, PortUse::New)` we should not try to retrieve an `eligible_listener` (which is binded to a local port)  but we should systematically use a new endpoint dedicated to dials without port_reuse

Reply:
Should be addressed in latest commit